ci: fix cla and dependabot notifications jobs

matifali opened the pull request:
When we merge a PR programmatically using `gh pr merge --auto --squash,` the Github auto handles the merge after all checks are passed. So, a `pull_request` event does not get triggered, and a merge commit is pushed to `main` by `github-actions[bot].` This change handles that case.

I also extracted all dependable logic to a separate workflow for better control and restored the original state of `contrib.yaml` so that cla can work as expected for Community PRs.

matifali committed:
ci: fix cla and dependabot notifications

matifali commented on .github/workflows/contrib.yaml:
```diff
@@ -2,19 +2,24 @@ name: contrib
       - synchronize
       - labeled
       - unlabeled
-      - opened
```
removed the duplicate from L8

matifali committed:
extract dependabot logic to dependabot.yaml

matifali committed:
remove dependabot logic

matifali committed:
Remove comment from pull_request_target section

ethanndickson commented on .github/workflows/contrib.yaml:
```diff
@@ -2,15 +2,14 @@ name: contrib
-    types: [created]
-  pull_request:
+    types: [created, edited]
+  pull_request_target:
```
I can't review it directly but `release-labels` only runs on `pull_request` still, so it currently won't fire. You could either move it to `dependabot.yaml` or make it run on `pull_request_target`.

matifali replied:
Thanks, Ethan. Fixing now.

dannykopping commented on .github/workflows/dependabot.yaml:
```diff
@@ -0,0 +1,91 @@
+  # Dependabot is annoying, but this makes it a bit less so.
+  dependabot-automerge:
+    runs-on: ubuntu-latest
+    if: github.event_name == 'pull_request' && github.event.pull_request.user.login == 'dependabot[bot]' && github.repository == 'coder/coder'
```
I still think we should match on the ID here instead of the name.

matifali replied:
I am not sure if that will work. But I can try. 
This is from GitHub docs where they recommend using this context for doing this workflow.

dannykopping commented on .github/workflows/dependabot.yaml:
```diff
@@ -0,0 +1,91 @@
+  dependabot-automerge-notify:
+    # Send a slack notification when a dependabot PR is merged.
+    runs-on: ubuntu-latest
+    if: github.event_name == 'push' && github.actor == 'github-actions[bot]'
```
Ditto.

matifali replied:
Added an additional check for `user_id`
Got it from: https://api.github.com/users/github-actions[bot]

dannykopping commented on .github/workflows/dependabot.yaml:
```diff
@@ -0,0 +1,91 @@
+    if: github.event_name == 'pull_request' && github.event.pull_request.user.login == 'dependabot[bot]' && github.repository == 'coder/coder'
+    permissions:
+      pull-requests: write
+      contents: write
```
Do we need this given that `pull-requests: write` is granted?

matifali replied:
Yes. It's needed to approve and merge PRs

dannykopping replied:
Got a reference for that?

matifali replied:
https://docs.github.com/en/code-security/dependabot/working-with-dependabot/automating-dependabot-with-github-actions#enabling-automerge-on-a-pull-request

I have also tested by removing these permissions and adding them as required. The merge action was failing without them.

matifali replied:
Added an additional check for `user_id`
Got it from: https://api.github.com/users/dependabot[bot].

dannykopping replied:
Sweet thanks 👍

dannykopping commented on .github/workflows/dependabot.yaml:
```diff
@@ -0,0 +1,91 @@
+  dependabot-automerge-notify:
+    # Send a slack notification when a dependabot PR is merged.
+    runs-on: ubuntu-latest
+    if: github.event_name == 'push' && github.actor == 'github-actions[bot]'
```
Wouldn't this trigger for any push by the github-actions bot? Can we not make this more specific?

matifali replied:
Yes but only the push to `main` branch. That is the same as merging a PR into `main`
GitHub actions bot don't push to `main` in other cases.

dannykopping commented on .github/workflows/dependabot.yaml:
```diff
@@ -0,0 +1,91 @@
+          github-token: "${{ secrets.GITHUB_TOKEN }}"
+
+      - name: Approve the PR
+        run: gh pr review --approve "$PR_URL"
```
Should we include a message here saying it was auto-approved?

matifali replied:
A message like a comment on the PR? I think it's not needed as the GitHub UI is self explanatory and it shows the PR is approved by a bot. 
We are additionally sending notifications for these too.

matifali replied:
I can add an `echo` statement here to print this ti logs.

matifali committed:
Update GitHub Actions PR validation logic

Switch `release-labels` job trigger to `pull_request_target` and refine
`dependabot` automerge conditions by verifying `actor_id`. Enhance
logging for approval and merge steps to provide better visibility.
